miredo/doc/miredo.conf.5
.\" ***********************************************************************
.\" *  Copyright © 2004-2006 Rémi Denis-Courmont.                         *
.\" *  This program is free software; you can redistribute and/or modify  *
.\" *  it under the terms of the GNU General Public License as published  *
.\" *  by the Free Software Foundation; version 2 of the license.         *
.\" *                                                                     *
.\" *  This program is distributed in the hope that it will be useful,    *
.\" *  but WITHOUT ANY WARRANTY; without even the implied warranty of     *
.\" *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               *
.\" *  See the GNU General Public License for more details.               *
.\" *                                                                     *
.\" *  You should have received a copy of the GNU General Public License  *
.\" *  along with this program; if not, you can get it from:              *
.\" *  http://www.gnu.org/copyleft/gpl.html                               *
.\" ***********************************************************************
.TH "MIREDO.CONF" "5" "$Date: 2006-10-26 12:58:33 -0600 (Thu, 26 Oct 2006) $" "miredo" "System Manager's Manual"
.SH NAME
miredo.conf \- configuration for Miredo
.SH SYNOPSIS
.B miredo.conf

.SH DESCRIPTON
.I miredo.conf
.RB " is the configuration file for " "Miredo" "."
Each configuration directive consists of one directive name, possibly
followed by one or several spaces and a value for the directive.
White spaces, empty lines and lines starting with '#' are ignored.

Directives are case-insensitive. A comprehensive list of the supported
directives follows:

.SH MODES

.TP
.BR "RelayType" " (" "client" "|" "relay" "|" "autoclient" ")"
Specifies what type of Teredo relay/client Miredo will serve as.
There are three possible choices:

.TP
.BR "client" " mode (the default)"
.RB "In " "client" " mode, " "Miredo" " acts as a "
.IR "Teredo client" ". Miredo will first authenticate with a "
.I "Teredo server"
.RB "(see " "ServerAddress" "), and if successful, will setup a Teredo"
tunneling interface with a public Teredo IPv6 address and a default
IPv6 route. That virtual networking interface can be used to reach the
IPv6 Internet as well as Teredo clients.

The use of Miredo as a Teredo client allows nodes to get an IPv6
connectivity from behind a NAT router device, as it tunnels IPv6
packets over UDP/IPv4 with special support for NAT routers. Routers of
that kind usually only forward TCP, UDP, and some ICMP, IPv4 packets
(with some limitation).

.BR "NOTE:" " Use of " "Miredo" " as a"
.IR "Teredo client" " is possible even if the system already has IPv6"
connectivity through another way (native, 6to4, TSP, etc). In that
case, Miredo will only be used to reach other Teredo clients, as its
tunneling interface has a default route with a higher metric (which is
to say a lower priority) than the other network interfaces.

.TP
.BR "autoclient" " mode"
.RB "In " "autoclient" " mode, " "Miredo" " acts as a "
.IR "Teredo client" " provided there is no other IPv6-enabled network"
interfaces. Whenever it detects the availability of global IPv6
connectivity through another interface, it stops its Teredo tunnel.

.TP
.BR "relay" " mode"
.RB "In " "relay" " mode, " "Miredo" " acts as a "
.IR "Teredo relay" ". It will assume that it has public global " "IPv4"
.RI "connectivity or is located behind a " "cone" " NAT device. "
Miredo will create a virtual networking interface with a route toward
Teredo clients.

.IR "Teredo relays" " forward IPv6 packets between " "Teredo clients"
.RB "and the IPv6 Internet. For that to work, Teredo relays " "MUST"
have a working IPv6 connectivity through a way distinct from Teredo
tunneling (native, 6to4, ISATAP, etc).

Warning: This mode should only be used if the node has a public IPv4
address or if it is behind a full cone NAT-router, otherwise the
tunnel will not work properly.

.RI "Miredo also accepts " "restricted" " and " "cone" " as aliases"
for the relay mode, for backward compatibility with older versions.

.SH CLIENT OPTIONS
.RI "The following directives are only available in " "(auto)client"
mode.

.TP
.BI "ServerAddress " "hostname"
.RB "The " "ServerAddress" " directive specifies the hostname or "
.RI "numerical IPv4 address of the " "Teredo server" " to use."
.IR "Teredo clients" " needs a  " "Teredo server" " to establish and"
maintain their IPv6 over UDP/IPv4 tunnel across a NAT device.

.RB "This directive " "MUST" " be specified when Miredo is in"
.IR "client" " mode. " "hostname" " must resolve to a valid IPv4"
address. If it is not present, and no server hostname is specified on
the command line when starting miredo either, the program will fail.

.TP
.BI "ServerAddress2 " "hostname2"
Miredo assumes that the secondary Teredo server address equals the
primary server address plus one. If that is not the case, this
directive must be used.

.SH RELAY OPTIONS
.RI "The following directive is only available in " "relay" " mode."
.RI "They are not available in " "(auto)client" " mode."

.TP
.BI "Prefix " "teredo_prefix"
.RB "This directive specifies the " "Teredo prefix" " which the Teredo"
relay and/or server will advertise.
.IR "teredo_prefix" " must be a valid IPv6 prefix."

.RB "The default value is " "2001:0000::" "."

.BR "Do not use that directive" " if you don't know what you are"
doing, as it is more than likely to break your Teredo connectivity.
That option must not be used when Miredo serves as a Teredo client.

.TP
.BI "InterfaceMTU " "mtu"
This directive overrides the default MTU size of 1280 bytes for the
Teredo tunneling interface. It should not be used if the default Teredo
prefix is used.

.SH GENERAL OPTIONS
.TP
.BI "InterfaceName " "ifname"
Specify the name of the Teredo tunneling interface which Miredo will
create ("miredo" by default). On some systems, it is not possible to
redefine the tunnel name.

.TP
.BI "BindAddress " "bind_address"
Bind the Teredo relay or Teredo client to a specific IPv4 address.
By default, it is not bound to any particular IPv4 address.

Use this option if you have trouble with the default value, such as if
you have a multi-homed host with equal-cost IPv4 routing, or if you
have specific firewalling constraints.

.TP
.BI "BindPort " "udp_port"
Define the UDP (IPv4) port number to be used by the relay or client.
By default, the operating system allocates an unused port
automatically.

Use this option if you have firewalling constraints which can cause
Miredo to fail when not using a fixed predefined port.

.TP
.BI "SyslogFacility " "facility"
Specify which syslog's facility is to be used by Miredo for logging.
.RB "Possible values are: " "daemon" " (the default), " "local0" ","
.RB "... " "local7" ", " "kern" " and " "user" " (see syslog(2))."

.SH "SEE ALSO"
miredo(8)

.SH AUTHOR
R\[char233]mi Denis-Courmont <rdenis at simphalempin.com>

$Id: miredo.conf.5 1805 2006-10-26 18:58:33Z remi $

http://www.simphalempin.com/dev/miredo/



miredo/NEWS
RELEASE NOTES for Miredo
stable release version 1.0.6
=============================
$Id: NEWS 1897 2006-12-31 11:19:27Z remi $

  This is a stable release. There are no known outstanding issues. We
strongly encourage users from previous releases to upgrade, as some important
fixes are included.
  Please report any problem by email to miredo-devel (at) remlab (dot) net

  You can find major changes since version 0.1.0 below. For more general
informations on this program, see README. For syntax details and
changes, run “miredo -h” to get help. For detailled internal changes, see
ChangeLog.

===========================================================================
STABLE RELEASE 1.0.6 : Major bug fixes

# Fix support for Teredo peers behind symmetric NATs (bug from 0.9.8).
  This could probably be used to spoof a Teredo clients.
# MacOS X compilation fix.

===========================================================================
STABLE RELEASE 1.0.5 : Major bug fixes

# Fix issues with timer and compiler optimizations (bug from 0.9.4).
# Increase ping test hop limit:
  Some IPv6 native nodes could not be reached otherwise.
# Increase HMAC secret size from 64 to 128 bits.
# Fix filtering of packets with link-local source address (bug from 0.4.0).
# Various fixes to the experimental ISATAP daemon.

============================================================================
STABLE RELEASE 1.0.4 : Major bug fix

# Fix packet storm with Teredo client behind symmetric NATs.

============================================================================
STABLE RELEASE 1.0.3 : Minor feature enhancement

# Work-around for compatibility with Microsoft Teredo servers.
# Better detection of symmetric NATs and access network changes.
# Limit spamming of syslog.

============================================================================
STABLE RELEASE 1.0.2 : Minor portability and bug fixes

# Several portability and minor/impossible bug fixes (see ChangeLog).

============================================================================
STABLE RELEASE 1.0.1 : Minor portability fixes

# Install configuration file samples into a dedicated directory.
# Build fixes (libteredo would not link on Mac OS X).

============================================================================
STABLE RELEASE 1.0.0 : Minor portability fixes

# Various minor build fixes.

============================================================================
RELEASE CANDIDATE 0.9.9 : Minor portability fixes

# NetBSD 4 build fixes.

============================================================================
RELEASE CANDIDATE 0.9.8 : Major feature enhancements, major security fixes

# Support interacting with Teredo clients behind symmetric NATs.
# Fix multiple problems with HMAC/ping authentication of non-Teredo nodes.
# Increase timestamp wrap time from 18 hours to 70 years to avoid replay
  attacks against authentication tokens.
# Allow Router Advertisement through the ISATAP tunnel (untested).
  isatapd remain very experimental, particular client-side.
  Someone ought to make a clean kernel implementation instead.
# Use HMAC instead of random nonces in Teredo clients.

=============================================================================
BETA RELEASE 0.9.7 : Major security fixes

# Fix infinite UDP packet forwarding loop in Teredo server (MTFL-SA 0603).

=============================================================================
BETA RELEASE 0.9.6 : Major compatibility fixes

# Lots of portability fixes, mostly for FreeBSD and the likes.
# Removed some dead code.
# Provide teredo-mire by default.
# Print more helpful error message for some common BSD tunneling issues.

=============================================================================
BETA RELEASE 0.9.5 : Major features enhancement, major bug fixes

# Removed too brittle cone NAT support. As a side effect, miredo is much
  faster to startup in client mode in most cases.
# Fix server-side handling of Windows Vista client solicitations.
# Removed the IgnoreConeBit configure option. The cone bit is now always
  ignored (this was the default ever since the introduction of that option).

=============================================================================
BETA RELEASE 0.9.4 : Major features enhancement

# More refined system clock usage brings about 30% performance boost.

=============================================================================
BETA RELEASE 0.9.3 : Major features enhancement, major bug fixes

# Use dedicated thread for packets transmission and reception. Miredo
  should now leverage dual-process, dual-core and SMT systems.

# Receive to-be-decapsulated and to-be-encapsulated packets in blocking
  mode; this improves performance by about 10% on Linux.

# Use the POSIX monotonic clock for maintenance procedure and “exclusive”
  mode watch if POSIX monotonic clock and clock selection are available.

# Minor optimizations to the most stressed code paths.
# Suppress spurious 4-seconds delay when waiting for “symmetric” probes.
# Minimalistic support for ISATAP client in ISATAPd.
# Ignore invalid Router Advertisements properly.
# Drop incoming multicast traffic as a precautionary measure.
# Truncate PID file properly when updating it.
# Fix deadlocks and spurious exits upon some signals (such as SIGCONT).
# Fix deadlock when Teredo server DNS hostname resolution fails.
# Handle would-be “spurious wakeups” properly.
# Fix IPv4 global unicast access-list.
# Fix SO_REUSEADDR socket option usage.
# Use C99 restrict keyword at sensible places for compiler optimization.
# Portability fixes for DragonFly BSD.
# Work-around for some uClibc POSIX defines insanity.
# Got rid of all C++ code.
# Renamed libteredo-mire to teredo-mire for consistency.
# Leverage newer autoconf macros.
# Fix encoding of non-ASCII characters in manual pages.
# Work-around for platforms that can't rename tunnel network interface.

miredo/src/miredo.c
/*
 * miredo.c - Miredo common daemon functions
 * $Id: miredo.c 1806 2006-10-26 19:22:45Z remi $
 *
 * See "Teredo: Tunneling IPv6 over UDP through NATs"
 * for more information
 */

/***********************************************************************
 *  Copyright © 2004-2006 Rémi Denis-Courmont.                         *
 *  This program is free software; you can redistribute and/or modify  *
 *  it under the terms of the GNU General Public License as published  *
 *  by the Free Software Foundation; version 2 of the license.         *
 *                                                                     *
 *  This program is distributed in the hope that it will be useful,    *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of     *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               *
 *  See the GNU General Public License for more details.               *
 *                                                                     *
 *  You should have received a copy of the GNU General Public License  *
 *  along with this program; if not, you can get it from:              *
 *  http://www.gnu.org/copyleft/gpl.html                               *
 ***********************************************************************/

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <gettext.h>

#include <string.h> // memset(), strsignal()
#include <stdlib.h> // exit()
#include <inttypes.h>
#include <stdbool.h>
#include <stdarg.h>

#include <pthread.h> // pthread_sigmask()
#include <signal.h> // sigaction()
#include <errno.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <syslog.h>
#include <unistd.h> // uid_t
#include <sys/wait.h> // waitpid()
#ifdef HAVE_SYS_CAPABILITY_H
# include <sys/capability.h>
#endif

#ifndef LOG_PERROR
# define LOG_PERROR 0
#endif

#include "miredo.h"
#include "conf.h"

/*
 * Signal handlers
 *
 * We block all signals when one of those we catch is being handled.
 * SECURITY NOTE: these signal handlers might be called as root or not,
 * in the context of the privileged child process or in that of the main
 * unprivileged worker process. They must not compromise the child's security.
 */
static volatile int should_exit;
static volatile int should_reload;

static void
exit_handler (int signum)
{
	if (should_exit)
		return;

	should_exit = signum;
	kill (0, signum);
}


static void
reload_handler (int signum)
{
	if (should_reload)
		return;

	should_reload = signum;
	kill (0, signum);
}


uid_t unpriv_uid = 0;
const char *miredo_chrootdir = NULL;

extern int
drop_privileges (void)
{
	/*
	 * We could chroot earlier, but we do it know to keep compatibility with
	 * grsecurity Linux kernel patch that automatically removes capabilities
	 * when chrooted.
	 */
	if ((miredo_chrootdir != NULL)
	 && (chroot (miredo_chrootdir) || chdir ("/")))
	{
		syslog (LOG_ALERT, _("Error (%s): %s\n"),
				"chroot", strerror (errno));
		return -1;
	}

	// Definitely drops privileges
	if (setuid (unpriv_uid))
	{
		syslog (LOG_ALERT, _("Error (%s): %s\n"), "setuid", strerror (errno));
		return -1;
	}

#ifdef HAVE_LIBCAP
	cap_t s = cap_init ();
	if (s != NULL)
	{
		cap_set_proc (s);
		cap_free (s);
	}
#endif
	return 0;
}


/**
 * Defines signal handlers.
 */
static void
InitSignals (void)
{
	struct sigaction sa;
	sigset_t set;

	should_exit = 0;
	should_reload = 0;

	memset (&sa, 0, sizeof (sa));
	sigemptyset (&sa.sa_mask); // -- can that really fail ?!?
	sigemptyset (&set);

	/* Signals that trigger a clean exit */
	sa.sa_handler = exit_handler;

	sigaction (SIGINT, &sa, NULL);
	sigaddset (&set, SIGINT);

	sigaction (SIGQUIT, &sa, NULL);
	sigaddset (&set, SIGQUIT);

	sigaction (SIGTERM, &sa, NULL);
	sigaddset (&set, SIGTERM);

	/* Signals that are ignored */
	sa.sa_handler = SIG_IGN;

	sigaction (SIGPIPE, &sa, NULL); // We check errno == EPIPE instead
	sigaddset (&set, SIGPIPE);

#if 0
	// might use these for other purpose in later versions:
	// or maybe not, GNU/kFreeBSD use these for pthread
	sigaction (SIGUSR1, &sa, NULL);
	sigaddset (&set, SIGUSR1);

	sigaction (SIGUSR2, &sa, NULL);
	sigaddset (&set, SIGUSR2);
#endif

	/* Signals that trigger a configuration reload */
	sa.sa_handler = reload_handler;

	sigaction (SIGHUP, &sa, NULL);
	sigaddset (&set, SIGHUP);

	/* Block all handled signals */
	pthread_sigmask (SIG_BLOCK, &set, NULL);
}


/*
 * Configuration and respawning stuff
 */
static void logger (void *dummy, bool error, const char *fmt, va_list ap)
{
	(void)dummy;

	vsyslog (error ? LOG_ERR : LOG_WARNING, fmt, ap);
}


extern int
miredo (const char *confpath, const char *server_name, int pidfd)
{
	int facility = LOG_DAEMON, retval;
	openlog (miredo_name, LOG_PID | LOG_PERROR, facility);

	miredo_conf *cnf = miredo_conf_create (logger, NULL);
	if (cnf == NULL)
		return -1;

	do
	{
		retval = 1;

		InitSignals ();

		if (!miredo_conf_read_file (cnf, confpath))
		{
			syslog (LOG_WARNING, _("Loading configuration from %s failed"),
			        confpath);
		}

		int newfac = LOG_DAEMON;
		miredo_conf_parse_syslog_facility (cnf, "SyslogFacility", &newfac);

		// Apply syslog facility change if needed
		if (newfac != facility)
		{
			closelog ();
			facility = newfac;
			openlog (miredo_name, LOG_PID, facility);
		}
		syslog (LOG_INFO, _("Starting..."));

		// Starts the main miredo process
		pid_t pid = fork ();

		switch (pid)
		{
			case -1:
				syslog (LOG_ALERT, _("Error (%s): %s\n"), "fork",
				        strerror (errno));
				break;

			case 0:
				close (pidfd);
				retval = miredo_run (cnf, server_name);
				miredo_conf_destroy (cnf);
				break;

			default:
				miredo_conf_clear (cnf, 0);
		}

		switch (pid)
		{
			case -1:
				continue;

			case 0:
				closelog ();
				exit (-retval);
		}

		sigset_t set, saved_set;
		sigemptyset (&set);
		pthread_sigmask (SIG_SETMASK, &set, &saved_set);

		// Waits until the miredo process terminates
		while (waitpid (pid, &retval, 0) == -1);

		pthread_sigmask (SIG_SETMASK, &saved_set, NULL);

		if (should_exit)
		{
			syslog (LOG_NOTICE, _("Exiting on signal %d (%s)"),
			        should_exit, strsignal (should_exit));
			retval = 0;
		}
		else
		if (should_reload)
		{
			syslog (LOG_NOTICE,
			        _("Reloading configuration on signal %d (%s)"),
			        should_reload, strsignal (should_reload));
			retval = 2;
		}
		else
		if (WIFEXITED (retval))
		{
			retval = WEXITSTATUS (retval);
			syslog (LOG_NOTICE, _("Terminated (exit code: %d)"),
			        retval);
			retval = retval != 0;
		}
		else
		if (WIFSIGNALED (retval))
		{
			retval = WTERMSIG (retval);
			syslog (LOG_INFO, _("Child %d killed by signal %d (%s)"),
			        (int)pid, retval, strsignal (retval));
			retval = 2;
		}
	}
	while (retval == 2);

	if (retval)
		syslog (LOG_INFO, _("Terminated with error(s)."));
	else
		syslog (LOG_INFO, _("Terminated with no error."));

	miredo_conf_destroy (cnf);
	closelog ();
	return -retval;
}


int (*miredo_diagnose) (void);
int (*miredo_run) (miredo_conf *conf, const char *server);

const char *miredo_name;

# ifdef HAVE_LIBCAP
const cap_value_t *miredo_capv;
int miredo_capc;
# endif

